lib/Conversion/CombToSMT/CombToSMT.cpp

```diff
@@ -0,0 +1,245 @@
+/// Lower the SourceOp to the TargetOp one-to-one.
+template <typename SourceOp, typename TargetOp>
+struct OneToOneOpConversion : OpConversionPattern<SourceOp> {
+  using OpConversionPattern<SourceOp>::OpConversionPattern;
+  using OpAdaptor = typename SourceOp::Adaptor;
+
+  LogicalResult
+  matchAndRewrite(SourceOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    rewriter.replaceOpWithNewOp<TargetOp>(
+        op,
+        OpConversionPattern<SourceOp>::typeConverter->convertType(
+            op.getResult().getType()),
+        adaptor.getOperands());
+    return success();
+  }
+};
+
+/// Converts an operation with a variadic number of operands to a chain of
+/// binary operations assuming left-associativity of the operation.
+template <typename SourceOp, typename TargetOp>
+struct VariadicToBinaryOpConversion : OpConversionPattern<SourceOp> {
+  using OpConversionPattern<SourceOp>::OpConversionPattern;
+  using OpAdaptor = typename SourceOp::Adaptor;
+
+  LogicalResult
+  matchAndRewrite(SourceOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    ValueRange operands = adaptor.getOperands();
+    if (operands.size() < 2)
+      return failure();
+
+    Value runner = operands[0];
+    for (Value operand : operands.drop_front())
+      runner = rewriter.create<TargetOp>(op.getLoc(), runner, operand);
+
+    rewriter.replaceOp(op, runner);
+    return success();
+  }
+};
+
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// Convert Comb to SMT pass
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct ConvertCombToSMTPass
+    : public impl::ConvertCombToSMTBase<ConvertCombToSMTPass> {
+  void runOnOperation() override;
+};
+} // namespace
+
+void circt::populateCombToSMTConversionPatterns(TypeConverter &converter,
+                                                RewritePatternSet &patterns) {
+  patterns.add<CombReplicateOpConversion, IcmpOpConversion, ExtractOpConversion,
+               SubOpConversion, MuxOpConversion,
+               OneToOneOpConversion<ShlOp, smt::BVShlOp>,
+               OneToOneOpConversion<ShrUOp, smt::BVLShrOp>,
+               OneToOneOpConversion<ShrSOp, smt::BVAShrOp>,
+               OneToOneOpConversion<DivSOp, smt::BVSDivOp>,
+               OneToOneOpConversion<DivUOp, smt::BVUDivOp>,
+               OneToOneOpConversion<ModSOp, smt::BVSRemOp>,
+               OneToOneOpConversion<ModUOp, smt::BVURemOp>,
+               VariadicToBinaryOpConversion<ConcatOp, smt::ConcatOp>,
+               VariadicToBinaryOpConversion<AddOp, smt::BVAddOp>,
+               VariadicToBinaryOpConversion<MulOp, smt::BVMulOp>,
+               VariadicToBinaryOpConversion<AndOp, smt::BVAndOp>,
+               VariadicToBinaryOpConversion<OrOp, smt::BVOrOp>,
+               VariadicToBinaryOpConversion<XorOp, smt::BVXOrOp>>(
```

Comment:
That's a very neat and compact way of getting lots of conversion patterns with similar behavior going. 💯